File: app/src/main/java/com/daohoangson/ydls/viewmodel/ActionLiveData.java
package com.daohoangson.ydls.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;

class ActionLiveData<T> extends LiveData<T> {
    void setActionData(T data) {
        setValue(data);
    }

    @Override
    public void observe(@NonNull LifecycleOwner owner, @NonNull final Observer<? super T> observer) {
        if (hasObservers()) {
            throw new RuntimeException("Too many observers");
        }

        super.observe(owner, new Observer<T>() {
            @Override
            public void onChanged(T data) {
                if (data == null) {
                    return;
                }

                observer.onChanged(data);
                setValue(null);
            }
        });
    }
}

File: app/src/main/java/com/daohoangson/ydls/viewmodel/ActionLiveData.kt
package com.daohoangson.ydls.viewmodel

import androidx.annotation.NonNull
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer

internal class ActionLiveData<T> : LiveData<T>() {
    fun setActionData(data: T) {
        value = data
    }

    override fun observe(@NonNull owner: LifecycleOwner, @NonNull observer: Observer<in T>) {
        if (hasObservers()) {
            throw RuntimeException("Too many observers")
        }

        super.observe(owner, Observer { data ->
            if (data == null) {
                return@Observer
            }

            observer.onChanged(data)
            value = null
        })
    }
}
